### PR TITLE
feat: introduce SpaceCompletion controller

### PR DESCRIPTION
**Pull request by MatousJobanek**
Adds SpaceCompletion controller which will set TierName and TargetCluster if they are not set already.
To choose the correct TargetCluter it uses the same capacity manager as is used for UserSignup

paired with: https://github.com/codeready-toolchain/toolchain-e2e/pull/469

### CODE REVIEW

File: controllers/spacecompletion/space_completion_controller.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+	// Fetch the Space
+	space := &toolchainv1alpha1.Space{}
+	err := r.Client.Get(context.TODO(), types.NamespacedName{
+		Namespace: r.Namespace,
+		Name:      request.Name,
+	}, space)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("Space not found")
```
Comment (xcoulon) [inline controllers/spacecompletion/space_completion_controller.go:62]: the `IsNotFound` case could be tested

File: controllers/spacecompletion/space_completion_controller.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, errs.Wrap(err, "unable to get the current Space")
```
Comment (xcoulon) [inline controllers/spacecompletion/space_completion_controller.go:69]: this case could be tested, too

File: controllers/spacecompletion/space_completion_controller.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	}
+
+	if space.Spec.TargetCluster == "" {
+		targetCluster, err := capacity.GetOptimalTargetCluster("", space.Namespace, r.GetMemberClusters, r.Client)
+		if err != nil {
+			return false, errs.Wrapf(err, "unable to get the optimal target cluster")
+		}
+		if targetCluster == "" {
+			logger.Info("no cluster available")
+			return false, nil
```
Comment (xcoulon) [inline controllers/spacecompletion/space_completion_controller.go:103]: so in that case, the `Space` will remain in "pending" state until there is a new reconcile loop trigger by a chanfge on the resource. 
Is it what we want to do, or should be requeue with a delay until there is room on a cluster? 
Also, in such a case, should we have an alert? 🚨

File: controllers/spacecompletion/space_completion_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+		space := spacetest.NewSpace("without-fields",
+			spacetest.WithTierName(""))
+		r, req, cl := prepareReconcile(t, space, getMemberClusters)
+
+		// when
+		_, err := r.Reconcile(context.TODO(), req)
+
+		// then
+		require.NoError(t, err)
+		spacetest.AssertThatSpace(t, test.HostOperatorNs, space.Name, cl).
```
Comment (xcoulon) [inline controllers/spacecompletion/space_completion_controller_test.go:41]: could you add an assertion on the response, in particular, on the `requeue` field? (assuming we expect `false` at this point?)

Comment (xcoulon) [reply controllers/spacecompletion/space_completion_controller_test.go:41]: (same comment applies on all tests below)

File: controllers/spacecompletion/space_completion_controller.go (hunk 1 above)
Comment (MatousJobanek) [reply controllers/spacecompletion/space_completion_controller.go:62]: We don't test these two cases in 99% of all controllers, so I didn't see it important doing that here. 
But I can add test case for that.

File: controllers/spacecompletion/space_completion_controller.go (hunk 3 above)
Comment (MatousJobanek) [reply controllers/spacecompletion/space_completion_controller.go:103]: The controller also watches ToolchainStatus, so as soon as there is any change in the clusters, then the controller will be triggered for the oldest Space that is in pending state.
In addition, ToolchainStatus is updated every 5 seconds or so, which will trigger the reconcile as well.

File: controllers/spacecompletion/space_completion_controller_test.go (hunk 4 above)
Comment (MatousJobanek) [reply controllers/spacecompletion/space_completion_controller_test.go:41]: it's supposed to be false literally in all cases, so I followed the same approach as we use in all other controllers like that - I don't check that.

File: controllers/spacecompletion/space_completion_controller.go (hunk 3 above)
Comment (xcoulon) [reply controllers/spacecompletion/space_completion_controller.go:103]: ah ok, I missed that :/ It explains why the e2e tests passed, then!

File: controllers/spacecompletion/space_completion_controller.go (hunk 1 above)
Comment (MatousJobanek) [reply controllers/spacecompletion/space_completion_controller.go:62]: added c7a09ec8

File: controllers/spacecompletion/space_completion_controller.go (hunk 2 above)
Comment (MatousJobanek) [reply controllers/spacecompletion/space_completion_controller.go:69]: added c7a09ec8

**Review by xcoulon — APPROVED** [review]

**Pull request merged**